### ISSUE TITLE
Reduce Scatter Kernels Not Launching

### ISSUE THREAD

**Issue by SeanNijjar** [tenstorrent/tt-metal#8172]
I'm having issues with kernel dispatch in the reduce scatter op I'm bringing up. This op isn't mainlined yet.

The fix for this issue is not to get passing reduce scatter tests (but that would be pretty crazy if it happened to pass on the first attempt lol), but rather just to resolve the issue that is preventing the kernels from running on the chips.

As soon as the reduce_scatter kernels are seen running on the workers/erisc with fixes on main - even if the reduce scatter test itself otherwise hangs/fails - then we can close this issue.

# Reproduce:
- Obtain T3000 machine (Mine is at 800MHz but I doubt that matters)
- checkout branch: `snijjar/issue-8172`
- build: `make build tests/tt_metal tests/tt_eager -j32`
- run: `pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_reduce_scatter.py -k "test_reduce_scatter_post_commit"`

# Symptoms
Based on the run log, I see that I'm enqueueing programs that look like the reduce scatter.
```
Launching Operation: "tt::tt_metal::ReduceScatter" (device<Tensors>)
 Attributes:
 	scatter_dim = 3
 	num_links = 1
 	ring_size = 4
 	ring_index = 0
 	receiver_device_id = 1
 	sender_device_id = 3
 	output_mem_config = tt::tt_metal::MemoryConfig(memory_layout=TensorMemoryLayout::INTERLEAVED,buffer_type=BufferType::DRAM,shard_spec=std::nullopt)
...
ReduceScatter invoking 4 worker cores and 2 erisc cores
...
Created EnqueueProgramCommand (active_cores: 6 bypass_mode: false expected_workers_completed: 0)
EnqueueCommandType::ENQUEUE_PROGRAM complete
```
(The above sequence with `Created EnqueueProgramCommand (active_cores: 6 bypass_mode:` is repeated #chips times)

[run.log](https://github.com/tenstorrent/tt-metal/files/15224809/run.log)

## Watcher Log

Across all of the chips, I only see the dispatcher kernels running:
```
k_id[0]: blank
k_id[44]: tt_metal/impl/dispatch/kernels/eth_tunneler.cpp
k_id[45]: tt_metal/impl/dispatch/kernels/packet_demux.cpp
k_id[46]: tt_metal/impl/dispatch/kernels/cq_prefetch.cpp
k_id[47]: tt_metal/impl/dispatch/kernels/cq_dispatch.cpp
k_id[48]: tt_metal/impl/dispatch/kernels/packet_mux.cpp
Dump #12 completed at 184.019s
```

[watcher.issue.log](https://github.com/tenstorrent/tt-metal/files/15224803/watcher.issue.log)

**Comment by SeanNijjar** [comment]
Marking this P1 for now as I have a few other items on my list that I can look at in the mean-time. Ideally this is resolved by the end of this week/before next week

**Comment by SeanNijjar** [comment]
I forgot to put this in the description initially but the fix for this issue is not to get passing reduce scatter tests (but that would be pretty crazy if it happened to pass on the first attempt lol), but rather just to resolve the issue that is preventing the kernels from running on the chips.

As soon as the reduce_scatter kernels are seen running on the workers/erisc with fixes on main - even if the reduce scatter test itself otherwise hangs/fails - then we can close this issue.

**Comment by ubcheema** [comment]
Kernels fail to launch because dispatch is waiting for a write barrier.
Write barrier is when dispatch has to make sure all pending writes are complete before we can start issuing control signals for the data written earlier.
 
There are multicast writes targeting workers, x1y1 to x1y10 (mcast noc x/y 0x00412810) in the following two dispatch write packed command. But this spans a harvested core. On my system, Rows 7 and 11 are harvested
Harvested cores will not return ack which means for the following two multicasts, fw counter will step ahead of hw writes acked by 2.
This results in hang at a subsequent write barrier.

There is also a latent bug here in that Row 10 is for dispatch cores, but this multicast is writing to L1 of and active dispatch cores.
x1y10 which is part of the multicast write group is also the dispatch core for device 0.

```
x=01,y=10 0x00088000 => 0x00010105 (65797)                                                                                                            
x=01,y=10 0x00088004 => 0x00019e90 (106128)                                                                                                            
x=01,y=10 0x00088008 => 0x00000004 (4)                                                                                                               
x=01,y=10 0x0008800c => 0x00000000 (0)                                                                                                               
x=01,y=10 0x00088010 => 0x00412810 (4270096)                                                                                                           
x=01,y=10 0x00088014 => 0x00000008 (8)                                                                                                              
x=01,y=10 0x00088018 => 0x00000000 (0)                                                                                                              
x=01,y=10 0x0008801c => 0x00000000 (0)                                                                                                              
x=01,y=10 0x00088020 => 0x00000000 (0)


x=01,y=10 0x00089000 => 0x00010105 (65797)            
x=01,y=10 0x00089004 => 0x00019e80 (106112)           
x=01,y=10 0x00089008 => 0x00000004 (4)             
x=01,y=10 0x0008900c => 0x00000000 (0)             
x=01,y=10 0x00089010 => 0x00412810 (4270096)          
x=01,y=10 0x00089014 => 0x00000008 (8)             
x=01,y=10 0x00089018 => 0x00000000 (0)             
x=01,y=10 0x0008901c => 0x00000000 (0)             
x=01,y=10 0x00089020 => 0x00000000 (0)
```

**Comment by ubcheema** [comment]
The issue of missing write ack is not due to harvesting but due to dispatch core being part of the multicast group without proper flag for self mcast.

Dispatch core should not have been part of mcast. This is due to test bug. 
Test is creating a worker grid size > the max allowed y size of 7 cores.

Flipping the arguments as shown below resolves the hang in dispatcher:

```
CoreRangeSet select_worker_cores(ccl::CCLOpConfig const& op_config, uint32_t num_links) {
    switch (op_config.get_topology()) {
        case tt::tt_metal::ccl::Topology::Linear:
            // 4 per direction per link
            return CoreRangeSet({CoreCoord(num_links - 1, 6), CoreRange(CoreCoord(0, 0))});
        case tt::tt_metal::ccl::Topology::Ring:
            // 4 per direction per link
            return CoreRangeSet({CoreCoord(num_links - 1, 6), CoreRange(CoreCoord(0, 0))});
        default: TT_ASSERT(false, "Unsupported topology"); return CoreRangeSet({});
    };
}
```
I have confirmed that with above change, kernels get launched on worker Tensix and Ethernet cores.


Another issue in the test is that Input Tensor is used to get the device id. And `detail::get_device` always returns Device 0 even if the kernels are to be loaded to Devices 1, 2, 3.
This results in all EP commands and kernels going to Device 0.


```
run_operation.cpp

OutputTensors run_device_operation(
    std::optional<std::reference_wrapper<CommandQueue>> queue,
    const DeviceOperation<OutputTensors>& operation,
    const Tensors& input_tensors,
    const OptionalConstTensors& optional_input_tensors,
    const OptionalTensors& optional_output_tensors) {
    ZoneScopedN("TT_DNN_DEVICE_OP");
    uint32_t op_id = assign_id();
...
...
...

    uint32_t device_id = detail::get_device(input_tensors, optional_input_tensors)->id();
```

**Comment by SeanNijjar** [comment]
Going to close this now as I'm using https://github.com/tenstorrent/tt-metal/issues/5562 to track main reduce scatter work which is no longer blocked.
